oak/server/rust/oak_runtime/src/node/wasm.rs

```diff
@@ -0,0 +1,703 @@
+/// the host ABI function mapping between the runtime and a Wasm instance, and the handle mapping
+/// between the instance and the runtime `Channel`s.
+///
+/// Any handle from `readers` or `writers` is unique to all handles in both `readers` and
+/// `writers`.
+struct WasmInterface {
+    /// A reference to the `Runtime` is needed to execute the host ABI functions.
+    runtime: Arc<Runtime>,
+
+    /// Reader channel mappings to unique u64 handles
+    readers: HashMap<u64, ChannelReader>,
+    /// Writer channel mappings to unique u64 handles
+    writers: HashMap<u64, ChannelWriter>,
+
+    /// A reference to the memory used by the `wasmi` interpreter. Host ABI functions using Wasm
+    /// relative addresses will perform reads/writes against this reference.
+    memory: Option<wasmi::MemoryRef>,
+}
+
+impl WasmInterface {
+    /// Generate a randomized handle. Handles are random to prevent accidental dependency on
+    /// particular values or sequence. See https://github.com/project-oak/oak/pull/347
+    fn allocate_new_handle(&mut self, channel: ChannelEither) -> u64 {
+        loop {
+            let handle = rand::thread_rng().next_u64();
+
+            if self.readers.get(&handle).is_some() || self.writers.get(&handle).is_some() {
+                continue;
+            }
+
+            match channel {
+                ChannelEither::Reader(channel) => {
+                    self.readers.insert(handle, channel);
+                }
+                ChannelEither::Writer(channel) => {
+                    self.writers.insert(handle, channel);
+                }
+            }
+
+            return handle;
+        }
+    }
+
+    /// Helper function to get memory.
+    fn get_memory(&self) -> &wasmi::MemoryRef {
+        self.memory
+            .as_ref()
+            .expect("WasmInterface memory not attached!?")
+    }
+
+    /// Make sure a given address falls within the currently allocated range of memory.
+    fn validate_address_arg(&self, addr: usize) -> Result<(), OakStatus> {
+        let byte_size: wasmi::memory_units::Bytes = self.get_memory().current_size().into();
+
+        if byte_size < wasmi::memory_units::Bytes(addr) {
+            return Err(OakStatus::ERR_INVALID_ARGS);
+        }
+
+        Ok(())
+    }
+
+    /// TODO(#472): doc
+    pub fn new(runtime: &Arc<Runtime>, initial_handle: ChannelReader) -> WasmInterface {
+        WasmInterface {
+            runtime: runtime.clone(),
+            readers: {
+                let mut r = HashMap::new();
+                r.insert(1, initial_handle);
+                r
+            },
+            writers: HashMap::new(),
+            memory: None,
+        }
+    }
+
+    /// TODO(#472): doc
+    fn node_create(
+        &self,
+        name_ptr: u32,
+        name_length: usize,
+        entrypoint_ptr: u32,
+        entrypoint_length: usize,
+        initial_handle: u64,
+    ) -> Result<(), OakStatus> {
```

Comment:
The existing `oak_tests` host function code is set up to have one or more `debug!` statements for each API invocation and result; it would be nice (for debugging) to have something equivalent throughout here.

Reply:
These methods get invoked via `invoke_index` line 455ish and are prefixed with `debug!` calls, which I think is what you are asking. E.g.

~~~rust
        match index {
            NODE_CREATE => {
                let name_ptr: u32 = args.nth_checked(0)?;
                let name_length: u32 = args.nth_checked(1)?;
                let entrypoint_ptr: u32 = args.nth_checked(2)?;
                let entrypoint_length: u32 = args.nth_checked(3)?;
                let initial_handle: u64 = args.nth_checked(4)?;

                debug!(
                    "{:?}: node_create: {} {} {} {} {}",
                    std::thread::current().id(),
                    name_ptr,
                    name_length,
                    entrypoint_ptr,
                    entrypoint_length,
                    initial_handle
                );

                map_host_errors(self.node_create(
                    name_ptr,
                    name_length,
                    entrypoint_ptr,
                    entrypoint_length,
                    initial_handle,
                ))
~~~